### PR TITLE
introduce implementation for date-time support in normalization

### PR DESCRIPTION
**Pull request by subodh1810**
Issue : https://github.com/airbytehq/airbyte/issues/1006
We are adding support for two new data types in normalization. 
1. TIMESTAMP WITH TIMEZONE
2. DATE

If a connector specifies the ISO 8601 format in JSON schema for a column, Airbyte would create that column as TIMESTAMP WITH TIMEZONE or DATE based on the format specified.
Ref : https://json-schema.org/understanding-json-schema/reference/string.html#dates-and-times

Note : We can not support TIMESTAMP for MySQL because MySQL doesnt allow `CAST` operation to work with TIMESTAMP data type. 
```

mysql> select CAST('2013-09-05T10:10:02Z' AS DATETIME) as tt;
+---------------------+
| tt                  |
+---------------------+
| 2013-09-05 10:10:02 |
+---------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select CAST('2013-09-05T10:10:02Z' AS TIMESTAMP) as tt;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'TIMESTAMP) as tt' at line 1
```
As you can see above, when I try to cast '2013-09-05T10:10:02Z' as DATETIME, it works but loses the Timezone info but when I try to cast it to TIMESTAMP, mysql throws an exception

I decided not to use DATETIME as a substitute cause it can not hold timezone information.
I also tried using the function `STR_TO_DATE` to make it work but it also didnt happen. `STR_TO_DATE` converts a string to DATETIME. 
```
mysql> select STR_TO_DATE('2013-09-05T10:10:02Z','%Y-%m-%dT%TZ') as tt;
+---------------------+
| tt                  |
+---------------------+
| 2013-09-05 10:10:02 |
+---------------------+
1 row in set (0.02 sec)

mysql> CREATE TABLE new_tt (select STR_TO_DATE('2013-09-05T10:10:02Z','%Y-%m-%dT%TZ') as tt);
Query OK, 1 row affected (0.03 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> describe new_tt;
+-------+----------+------+-----+---------+-------+
| Field | Type     | Null | Key | Default | Extra |
+-------+----------+------+-----+---------+-------+
| tt    | datetime | YES  |     | NULL    |       |
+-------+----------+------+-----+---------+-------+
1 row in set (0.02 sec)

```
As you can see above, the value `2013-09-05T10:10:02Z` lost its TIMEZONE info after passing through `STR_TO_DATE` even when the format `'%Y-%m-%dT%TZ'` consisted of timezone info. Also the table created has the data type as DATETIME

Connectors would need to make changes in their catalog creation logic and now start attaching format option for columns ([example](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-harvest/source_harvest/schemas/billable_rates.json#L21)).

Note:
All the files under `airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output` are auto generated as a result of running the tests.
 The main logic change lives in 
1. `airbyte-integrations/bases/base-normalization/dbt-project-template/macros/cross_db_utils/datatypes.sql`
2. `airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by subodh1810** [comment]
@sherifnada I updated the description of the PR to highlight the auto generated files

> All the files under airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output are auto generated as a result of running the tests.
> The main logic change lives in
> 
> airbyte-integrations/bases/base-normalization/dbt-project-template/macros/cross_db_utils/datatypes.sql
> airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py
>

**Review by ChristopheDuong — APPROVED** [review]

**Comment by ChristopheDuong** [comment]
if connectors specify in the catalog columns with date/timestamp format but don't provide values in the ISO 8601 format in JSON schema for such columns, dbt would fail to run with invalid cast right?

**Review by sherifnada — APPROVED** [review]
one question about datetime but otherwise looks great 🎉

File: airbyte-integrations/bases/base-normalization/dbt-project-template/macros/cross_db_utils/datatypes.sql [hunk 1]
```diff
@@ -60,3 +60,33 @@
...
+{% macro default__type_timestamp_with_timezone() %}
+    timestamp with time zone
+{% endmacro %}
+
+{% macro bigquery__type_timestamp_with_timezone() %}
+    timestamp
+{% endmacro %}
+
+{%- macro mysql__type_timestamp_with_timezone() -%}
+    char
```
[the remark this answers is not in the document]
Comment (subodh1810) [reply airbyte-integrations/bases/base-normalization/dbt-project-template/macros/cross_db_utils/datatypes.sql:80]: As discussed over the call, using DATETIME for MySQL is a bit complicated cause it doesn't allow specifying a time zone offset before version 8.0.19. So its best to use char for now

**Comment by subodh1810** [comment]
@ChristopheDuong regarding this, Yes the cast would fail.
I and @sherifnada had a discussion and this problem is the same for other data types right now. Like if you specify the type as number and value for that column is string then the cast would fail in that case as well. There are few ways to handle it, we could put JSON schema validation in place for each record that we read from source to make sure the value is in sync with the JSON schema but it would have a performance impact. For now we decided we would continue with the same behaviour.


> if connectors specify in the catalog columns with date/timestamp format but don't provide values in the ISO 8601 format in JSON schema for such columns, dbt would fail to run with invalid cast right?

**Comment by subodh1810** [comment]
So just had a discussion with @marcosmarxm , he also has a PR for normalization changes https://github.com/airbytehq/airbyte/pull/4557 and we decided that we would publish both the changes together. I am just going to merge this PR and then Marcos will re-base his branch with my changes and then publish both the things together

**Pull request merged**